client/jetpack-cloud/sections/partner-portal/company-details-form/index.tsx

```diff
@@ -0,0 +1,193 @@
+import { Button, Card } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import { ReactElement, useCallback, useState } from 'react';
+import { useDispatch, useSelector } from 'react-redux';
+import FormFieldset from 'calypso/components/forms/form-fieldset';
+import FormLabel from 'calypso/components/forms/form-label';
+import FormTextInput from 'calypso/components/forms/form-text-input';
+import SelectDropdown from 'calypso/components/select-dropdown';
+import TextPlaceholder from 'calypso/jetpack-cloud/sections/partner-portal/text-placeholder';
+import { recordTracksEvent } from 'calypso/state/analytics/actions';
+import { errorNotice, removeNotice, successNotice } from 'calypso/state/notices/actions';
+import useUpdateCompanyDetailsMutation from 'calypso/state/partner-portal/partner/hooks/use-update-company-details';
+import { getCurrentPartner } from 'calypso/state/partner-portal/partner/selectors';
+import { APIError } from 'calypso/state/partner-portal/types';
+import { useCountriesAndStates } from './hooks/use-countries-and-states';
+import './style.scss';
+
+export default function CompanyDetailsForm(): ReactElement {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+	// The partner has already been fetched through e.g. the requireAccessContext
+	// controller function, so we do not have to do any further checks.
+	const partner = useSelector( getCurrentPartner );
+	const { countryOptions, stateOptionsMap } = useCountriesAndStates();
+	const submitNotificationId = 'company-details-form';
+	const showCountryFields = countryOptions.length > 0;
+
+	const [ isSubmitting, setIsSubmitting ] = useState( false );
+	const [ name, setName ] = useState( partner?.name ?? '' );
+	const [ country, setCountry ] = useState( partner?.address.country ?? '' );
+	const [ city, setCity ] = useState( partner?.address.city ?? '' );
+	const [ line1, setLine1 ] = useState( partner?.address.line1 ?? '' );
+	const [ line2, setLine2 ] = useState( partner?.address.line2 ?? '' );
+	const [ postalCode, setPostalCode ] = useState( partner?.address.postal_code ?? '' );
+	const [ addressState, setAddressState ] = useState( partner?.address.state ?? '' );
+
+	const stateOptions = stateOptionsMap.hasOwnProperty( country )
+		? stateOptionsMap[ country ]
+		: false;
+
+	const updateCompanyDetails = useUpdateCompanyDetailsMutation( {
+		onSuccess: () => {
+			dispatch(
+				successNotice( translate( 'Company details has been updated' ), {
+					id: submitNotificationId,
+				} )
+			);
+			setIsSubmitting( false );
+		},
+		onError: ( error: APIError ) => {
+			dispatch(
+				errorNotice( error.message, {
+					id: submitNotificationId,
+				} )
+			);
+			setIsSubmitting( false );
+		},
+	} );
+
+	const handleSubmit = useCallback(
+		( e ) => {
+			e.preventDefault();
```

Comment:
We should have an early return here to not submit the form if the country list or an update request are still pending:
```js
if ( ! showCountryFields || updateCompanyDetails.isLoading ) {
    return;
}
```